Particle Rotation Manipulation

AatreyuShau opened the pull request:
Allows the user to prevent particle Billboarding
and manipulate particle rotation
via a procedure that returns a vector
<img width="819" height="302" alt="Screenshot 2025-12-27 at 11 52 49 PM" src="https://github.com/user-attachments/assets/8cb09128-0aed-485e-91a5-19b51daad6a1" />

https://github.com/user-attachments/assets/6a7570d3-0199-4eff-8323-8fdde415c79e

Updated workspace: [particleTestWithWorldNew.zip](https://github.com/user-attachments/files/24361499/particleTestWithWorldNew.zip)

KlemenDEV commented on src/main/java/net/mcreator/element/types/Particle.java:
```diff
@@ -62,6 +62,9 @@
 	public boolean alwaysShow;
 	public boolean emissiveRendering;
 
+	public boolean lockRot;
```
1. Use more understandable names, lockRotation, for example
2. Specify default value. Existing elements should likely have this value at true

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -3091,6 +3091,8 @@ elementgui.particle.gravity=Particle gravity:
 elementgui.particle.always_show=Always show the particle?
 elementgui.particle.does_collide=Does particle collide?
+elementgui.particle.lock_rot=Does Particle billboard?
+elementgui.particle.rotation_quat=Particle Direction:
```
Non standard capitalization of letters

KlemenDEV reviewed (commented):
Tests seem to be failing. Did you run tests locally?

KlemenDEV commented:
I recommend running tests locally with debugger, it is much easer to work this way and not rely on guesswork to fix things

KlemenDEV commented on plugins/mcreator-localization/help/default/particle/lock_rot.md:
```diff
@@ -0,0 +1,3 @@
+This parameter controls if the particles billboards(always appears to face player).
```
"billboard ("

AatreyuShau commented on src/main/java/net/mcreator/element/types/Particle.java:
```diff
@@ -62,6 +62,9 @@
 	public boolean alwaysShow;
 	public boolean emissiveRendering;
 
+	public boolean lockRotation;
```
@KlemenDEV Yes particles billboard by default. 
This option currently disables billboarding.
Wouldn't making the checkbox 'enable billboarding', cause issues for existing workspaces?

KlemenDEV replied:
Your PR provided checkbox to enable it and it was off by default, that's why I was asking as this would break workspaces.

In this case, I guess you should just leave the procedure selector

KlemenDEV commented:
Why is the checkbox back?

AatreyuShau commented:
@KlemenDEV Everything should be sorted now. Tests are fine; I’ve removed the checkbox as well.
I also fixed a minor issue; the procedure should be considered client-side now
Please look into it.

KlemenDEV commented on plugins/mcreator-localization/help/default/particle/rotation_provider.md:
```diff
@@ -0,0 +1,7 @@
+This parameter controls if the particle billboards (always appears to face player)
```
> This parameter controls if the particle billboards (always appears to face player)

IIRC particles billboard by default, so it would be beter to say "this parameter overrides default billboarding (always appears to face player) behavior" or something like that

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/templates/particle.java.ftl:
```diff
@@ -111,6 +111,16 @@ package ${package}.client.particle;
 
+	<#if hasProcedure(data.rotationProvider)>
+	@Override public void render(VertexConsumer buffer, Camera camera, float ageTicks) {
+		Quaternionf spinX = new Quaternionf().rotationX((float) Math.toRadians((<@procedureOBJToConditionCode data.rotationProvider/>).x()));
```
Apples to both versions

KlemenDEV replied:
You should use rotationXYZ

AatreyuShau commented on plugins/mcreator-localization/help/default/particle/rotation_provider.md:
```diff
@@ -0,0 +1,7 @@
+where: `vec(pitch, yaw, roll)`
+- (first entry provides pitch, second one provides yaw, third gives roll)
+
+Note: When rendering, the rotations are applied in the order: (pitch-yaw-roll)
```
I've actually changed the order of multiplication itself so the statement here now stands true.

KlemenDEV replied:
I doubt rotateXYZ does PYR order since this is quite non standard @AatreyuShau

AatreyuShau replied:
I just checked the joml definition for it and have updated it accordingly

AatreyuShau commented:
@KlemenDEV I've made changes as per your request. Please look into it

AatreyuShau commented:
@KlemenDEV I've made the changes you requested and run the test. Everything should be sorted

KlemenDEV commented on src/main/java/net/mcreator/ui/modgui/ParticleGUI.java:
```diff
@@ -77,6 +78,10 @@ public ParticleGUI(MCreator mcreator, ModElement modElement, boolean editingMode
 				Dependency.fromString("x:number/y:number/z:number/world:world/age:number/scale:number"));
+		rotationProvider = new ProcedureSelector(this.withEntry("particle/rotation_provider"), mcreator,
+				L10N.t("elementgui.particle.rotation_provider"), ProcedureSelector.Side.CLIENT, true,
+				VariableTypeLoader.BuiltInTypes.VECTOR, Dependency.fromString("ageTicks:number")).setDefaultName(
```
Other procedures provide much more than just age, I suggest you follow format of other procedures for particle that provide much more dependencies particle exposes

KlemenDEV commented:
Please see https://github.com/MCreator/MCreator/pull/6017#discussion_r2649207239

AatreyuShau commented on plugins/generator-1.21.1/neoforge-1.21.1/templates/particle.java.ftl:
```diff
@@ -111,6 +111,23 @@ package ${package}.client.particle;
+	<#if hasProcedure(data.rotationProvider)>
+	@Override public void render(VertexConsumer buffer, Camera camera, float partialTicks) {
+		Level world = this.level;
+		double x = Mth.lerp(partialTicks, this.xo, this.x);
```
Actually that could cause unintended jittering this is the better implementation

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/templates/particle.java.ftl:
```diff
@@ -111,6 +111,31 @@ package ${package}.client.particle;
+		Quaternionf tilt = new Quaternionf().rotationXYZ((float) vec.x(), (float) vec.y(), (float) vec.z());
+		Quaternionf flippedTilt = new Quaternionf(tilt).mul(flip);
+		this.renderRotatedQuad(buffer, camera, tilt, partialTicks);
+		this.renderRotatedQuad(buffer, camera, flippedTilt, partialTicks); // render a flipped face because by default only a single side renders this makes particle visible from all angles
```
Also is there any example of other mods doing this or vanilla particles? Does this code follow standard modding practices?

KlemenDEV replied:
@AatreyuShau, if possible, please answer to my question

AatreyuShau commented:
@KlemenDEV I've updated the ftls as per your requests. Please look into it.
Also, should I provide the workspace zip here or should I dm it to you?

KlemenDEV commented:
Best you post it here if you are fine with it since other can then try it too :)

AatreyuShau commented:
@KlemenDEV Alright 
[particleTestWithWorld.zip](https://github.com/user-attachments/files/24359358/particleTestWithWorld.zip)

AatreyuShau commented:
@KlemenDEV I made a minor change in the dependencies and ran the tests.. It procedure should now provide the particles total interpolated age instead of just the partialTicks..
Please look into it.

KlemenDEV commented on plugins/generator-1.21.1/neoforge-1.21.1/templates/particle.java.ftl:
```diff
@@ -111,6 +111,29 @@ package ${package}.client.particle;
+	<#if hasProcedure(data.rotationProvider)>
+	@Override public void render(VertexConsumer buffer, Camera camera, float partialTicks) {
+		Quaternionf flip = new Quaternionf().rotateY((float)Math.PI);
+		double particleAge = this.age + partialTicks;
```
Inline expression in the procedure provider itself instead of "particleAge"

KlemenDEV reviewed (approved):
LGTM, thanks!
